File: ios/DesMoinesInsider/Services/SponsoredPickService.swift
import Foundation

// MARK: - IOS-ADS-015 · Sponsored picks in AI discovery flows
//
// Client for the `get-sponsored-pick` edge function. Surfaces AT MOST ONE
// clearly-labeled sponsored listing inside the high-intent native AI flows
// (Ask Pulse, Surprise Me, and — once it ships — Trip Planner / IOS-PARITY-001).
//
// Relevance + eligibility live SERVER-SIDE (the edge function), so the app only
// asks "is there a sponsored pick for this surface + query?" and renders the
// result with a visible "Sponsored" label. Free-tier only: premium users are
// short-circuited locally (no network) AND re-checked server-side.
//
// Impressions/clicks are tracked through `AdTrackingService` under the
// `sponsored_listing` inventory class, sharing the same analytics + dedupe as
// the in-feed sponsored listings (IOS-ADS-011 / IOS-ADS-014).
@MainActor
@Observable
final class SponsoredPickService {
    static let shared = SponsoredPickService()
    private init() {}

    enum Surface: String {
        case askPulse = "ask_pulse"
        case surpriseMe = "surprise_me"
        case tripPlanner = "trip_planner"
    }

    struct SponsoredPick: Identifiable, Decodable, Hashable {
        let itemType: String   // "event" | "restaurant"
        let itemId: String
        let title: String
        let reason: String
        let imageUrl: String?
        let campaignId: String

        var id: String { "\(itemType)-\(itemId)" }
    }

    private let supabase = SupabaseService.shared.client
    private let storeKit = StoreKitService.shared
    private let tracking = AdTrackingService.shared

    /// Returns a sponsored pick for the surface, or nil when none is eligible.
    /// Always nil for premium tiers and during UI tests.
    func pick(for surface: Surface, query: String? = nil) async -> SponsoredPick? {
        guard !Config.isUITesting else { return nil }
        // Free-tier only — short-circuit before hitting the network.
        guard storeKit.currentTier == .free else { return nil }
        guard let client = supabase else { return nil }

        struct Payload: Encodable {
            let surface: String
            let query: String?
        }
        struct Envelope: Decodable {
            let pick: SponsoredPick?
        }

        do {
            let envelope: Envelope = try await client.functions.invoke(
                "get-sponsored-pick",
                options: .init(body: Payload(surface: surface.rawValue, query: query)),
            )
            return envelope.pick
        } catch {
            #if DEBUG
            AppLogger.network.warning("SponsoredPick fetch failed: \(error.localizedDescription)")
            #endif
            return nil
        }
    }

    func logImpression(_ pick: SponsoredPick, surface: Surface) {
        tracking.logSponsoredImpression(
            listingType: pick.itemType,
            listingId: pick.itemId,
            placement: surface.rawValue,
        )
    }

    func logClick(_ pick: SponsoredPick, surface: Surface) {
        tracking.logSponsoredClick(
            listingType: pick.itemType,
            listingId: pick.itemId,
            placement: surface.rawValue,
        )
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/data/remote/SponsoredPickService.kt
package com.desmoines.aipulse.data.remote

import com.desmoines.aipulse.data.model.SponsoredPick
import com.desmoines.aipulse.data.model.SponsoredPickResponse
import com.desmoines.aipulse.util.AppLogger
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.functions.functions
import io.ktor.client.statement.bodyAsText
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Fetches a labeled sponsored pick for an AI surface via `get-sponsored-pick`.
 * Free tier only: premium tiers short-circuit to null locally (the server also
 * re-checks via the bearer token). Any failure / no inventory returns null.
 * Mirrors iOS SponsoredPickService.
 */
@Singleton
class SponsoredPickService @Inject constructor(
    private val supabaseClient: SupabaseClient?,
    private val billingService: BillingService,
    private val json: Json,
) {
    /** Valid AI surfaces accepted by the edge function. */
    object Surface {
        const val ASK_PULSE = "ask_pulse"
        const val SURPRISE_ME = "surprise_me"
        const val TRIP_PLANNER = "trip_planner"
    }

    suspend fun fetch(surface: String, query: String? = null): SponsoredPick? {
        // Premium users never see sponsored picks — skip the network entirely.
        if (billingService.currentTier.value.isPremium) return null
        val client = supabaseClient ?: return null

        return runCatching {
            val payload = buildJsonObject {
                put("surface", surface)
                query?.trim()?.takeIf { it.isNotEmpty() }?.let { put("query", it) }
            }
            val response = client.functions("get-sponsored-pick", body = payload)
            if (response.status.value !in 200..299) return@runCatching null
            json.decodeFromString(SponsoredPickResponse.serializer(), response.bodyAsText())
                .pick
                ?.takeIf { it.isRenderable }
        }.onFailure { AppLogger.network.warning("get-sponsored-pick failed: ${it.message}") }
            .getOrNull()
    }
}
